Optimize retrieving role variables

mkrizek opened the pull request:
##### SUMMARY
Optimize retrieving role variables

This patch optimizes performance of getting variables for a role by
changing two things:

1. In Role.get_default_vars() we called get_default_vars() recursively
   for all dependencies duplicating efforts. It is sufficient to just
   process _default_vars for all dependencies instead.

2. In Role.get_vars() we called dep.get_vars(include_params=False, only_exports=True)
   for all dependencies. Looking at what the call returns given the values
   of its arguments more closely it is just the dep's _role_vars. So
   replace the expensive call with dep._role_vars.

With these two changes the time to run a testing playbook with many
roles and variables goes down from:

35,19s user 4,15s system 100% cpu 39,185 total

to:

7,53s user 3,58s system 101% cpu 10,960 total

I also added a FIXME that would make the execution a bit more faster:

6,34s user 3,17s system 101% cpu 9,351 total

at the expense of the change being more invasive. Since the performance
improvement is very good with minimal changes already, we can do that
as a part of future changes.

ci_complete

##### ISSUE TYPE

<!--- Pick one below and delete the rest -->

- Bugfix Pull Request

webknjaz commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -115,9 +109,9 @@ def __init__(self,
-        self._role_name: str = None
-        self._role_path: str = None
-        self._role_collection: str = None
+        self._role_name: str | None = None
```
When there's `= None` in the initial assignment combined with a type annotation, `| None` is implied automatically, so it's typically unnecessary to clutter types with it (same in function annotations).

mkrizek replied:
I changed this because PyCharm complained. Reverted as it's not related to this PR anyway.

webknjaz replied:
Yeah, that would be type-checker dependent. I think it's configurable in MyPy but pycharm has their own thing in Java IIRC.

webknjaz replied:
UPD: Matt Clay said that we should always have a `| None` for when MyPy gets configured stricter.

webknjaz commented on lib/ansible/playbook/play.py:
```diff
@@ -89,7 +89,7 @@ def __init__(self):
         self._included_path = None
         self._removed_hosts = []
-        self.role_cache = {}
+        self.role_cache: dict[str, list[Role]] = {}
```
Looks like `__init__()` above needs a `-> None`.

webknjaz commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -141,11 +135,13 @@ def __init__(self,
         self._handler_blocks: list[Block] = []
         self._compiled_handler_blocks: list[Block] | None = None
         self._default_vars: dict[str, str] | None = dict()
+        self._deps_default_vars: dict[str, str] | None = dict()
+        self._deps_export_vars: dict[str, str] | None = dict()
```
AFAICS, neither can be `None` in runtime — the are both coming from a `combine_vars()` invocation that never return `None`.
```suggestion
        self._deps_default_vars: dict[str, str] = {}
        self._deps_export_vars: dict[str, str] = {}
```

webknjaz commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -472,7 +475,7 @@ def add_parent(self, parent_role):
             self._parents.append(parent_role)
 
     def get_parents(self):
```
Given that the patch changes the behavior a little, it might be a good idea to document the intent:
```suggestion
    def get_parents(self):
        """Return a copy of the parents roles list."""
```

mkrizek committed:
Optimize retrieving role variables

This patch optimizes performance of getting variables for a role by
changing two things:

1. In Role.get_default_vars() we called get_default_vars() recursively
   for all dependencies duplicating efforts. It is sufficient to just
   process _default_vars for all dependencies instead.

2. In Role.get_vars() we called dep.get_vars(include_params=False, only_exports=True)
   for all dependencies. Looking at what the call returns given the values
   of its arguments more closely it is just the dep's _role_vars. So
   replace the expensive call with dep._role_vars.

With these two changes the time to run a testing playbook with many
roles and variables goes down from:

35,19s user 4,15s system 100% cpu 39,185 total

to:

7,53s user 3,58s system 101% cpu 10,960 total

I also added a FIXME that would make the execution a bit more faster:

6,34s user 3,17s system 101% cpu 9,351 total

at the expense of the change being more invasive. Since the performance
improvement is very good with minimal changes already, we can do that
as a part of future changes.

ci_complete

s-hertel commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -496,7 +496,9 @@ def get_default_vars(self, dep_chain=None):
+        # FIXME combining deps' default_vars can be optimized by moving it
+        #       into _load_role_data so it is done just once
+        for dep in self._dependencies:
             default_vars = combine_vars(default_vars, dep.get_default_vars())
```
Can we use the same pattern for defaults and vars, just to make these methods more consistent? For example

```python
        for dep in self.get_all_dependencies():
            default_vars = combine_vars(default_vars, dep._default_vars)
```

or in get_vars maybe
```python
        for dep in self._dependencies:
            # Avoid rerunning dupe deps since they can have vars from previous invocations and they accumulate in deps
            # TODO: re-examine dep loading to see if we are somehow improperly adding the same dep too many times
            if dep not in seen:
                # only take 'exportable' vars from deps
                all_vars = combine_vars(all_vars, dep.get_vars(include_params=False, only_exports=True))
```

mkrizek replied:
Both options would be slower, so we would be trading consistency for performance in this case. I am fine with doing:
```python
        for dep in self.get_all_dependencies():
            default_vars = combine_vars(default_vars, dep._default_vars)
```

But the latter option, the `dep.get_vars()` call that only does part of the work depending on the arguments passed is unnecessarily expensive.

Or maybe I can just add comments to explain why it's inconsistent.

mkrizek replied:
FWIW I ended up making the calls consistent.

s-hertel reviewed (approved):
nice simplification
